substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/memory/UntrackedNullableNativeMemory.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2024, 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.core.memory;
+
+import static com.oracle.svm.core.Uninterruptible.CALLED_FROM_UNINTERRUPTIBLE_CODE;
+
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.impl.UnmanagedMemorySupport;
+import org.graalvm.word.PointerBase;
+import org.graalvm.word.UnsignedWord;
+import org.graalvm.word.WordFactory;
+
+import com.oracle.svm.core.Uninterruptible;
+
+import jdk.graal.compiler.api.replacements.Fold;
+
+/**
+ * Manages native memory. This class explicitly does <b>NOT</b> support native memory tracking
+ * (NMT). It can therefore be used for the following use cases:
+ * <ul>
+ * <li>Allocate native memory that is later freed outside of Native Image code (e.g., in a C
+ * library).</li>
+ * <li>Free native memory that was allocated outside of Native Image code (e.g., in a C
+ * library).</li>
+ * </ul>
+ */
+public class UntrackedNullableNativeMemory {
```

Comment:
Why have `UntrackedNullableNativeMemory` wrap `UnmanagedMemorySupport` if they serve a similar purpose? Is it a concern that a caller could either call `UnmanagedMemorySupport` or `UntrackedNullableNativeMemory` directly to accomplish the same thing? They could even use `LibC` directly, if they are determined.

Reply:
`UnmanagedMemory` and `UnmanagedMemorySupport` shouldn't be used internally. By not using it internally, it makes it easier to distinguish between code that bypasses NMT on purpose and code that accidentally uses the public API. Note that `UnmanagedMemorySupport` is platform-specific and only some implementations delegate to `LibC`.